src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnostics.cs

```diff
@@ -96,19 +97,11 @@ protected async Task AppendDiagnosticsAsync(Solution solution, CancellationToken
             {
                 // PERF: run projects in parallel rather than running CompilationWithAnalyzer with concurrency == true.
                 // We do this to not get into thread starvation causing hundreds of threads to be spawned.
-                var includeProjectNonLocalResult = true;
-
-                var tasks = new Task[solution.ProjectIds.Count];
-                var index = 0;
-                foreach (var project in solution.Projects)
-                {
-                    var localProject = project;
-                    tasks[index++] = Task.Run(
-                        () => AppendDiagnosticsAsync(
-                            localProject, localProject.DocumentIds, includeProjectNonLocalResult, cancellationToken), cancellationToken);
-                }
-
-                await Task.WhenAll(tasks).ConfigureAwait(false);
+                await RoslynParallel.ForEachAsync(
```

Comment:
>ForEachAsync [](http://example.com/codeflow?start=37&length=12)

Maybe another place where the code would be better served by producer/consumer? It might get rid of the lock in appenddiagnostics which would be nice.

Reply:
oh, I was looking at the wrong method, maybe this doesn't work